### PR TITLE
Support doctests in internal libraries and executables

### LINKED ISSUE #27: Build_doctests isn't generated unless Cabal file defines a library
I tried for quite a while now to run a doctest suite with `cabal-doctest` in a package that defines an executable, but no library, and all attempts invariably failed because the required `Build_doctests.hs` file wasn't generated. Then I added an *empty* `Library` stanza to my Cabal file, and immediately everything worked! I'm not sure why things work that way, but it doesn't seem right. Is that a bug, maybe?

### PR DESCRIPTION
**Pull request by RyanGlScott**
This augments `Build_doctests` with extra sets of flags for each named component (such as internal libraries and executables). Along the way, I added a separate example that demonstrates this capability (`multiple-components-example`), and renamed the existing example to `simple-example` for clarity.

Fixes #27.

### CODE REVIEW

File: src/Distribution/Extra/Doctest.hs [hunk 1]
```diff
@@ -304,6 +330,30 @@ generateBuildModule testSuiteName flags pkg lbi = do
...
+    executableName = unUnqualComponentName . exeName
+#else
+    executableName = exeName
+#endif
+
+    exeMainIs :: Executable -> [ModuleName]
+    exeMainIs exe = [ ModuleName.fromString
+                    $ dropExtension -- Be careful: the IsString instance doesn't
+                                    -- like "Main.hs", only "Main"
+                    $ modulePath exe
```
Comment (phadej) [inline src/Distribution/Extra/Doctest.hs]: executable file name may be `Something.hs`, when the module is still `Main`, for it we need to pass actual file name.

File: src/Distribution/Extra/Doctest.hs [hunk 2]
```diff
@@ -165,88 +172,107 @@ generateBuildModule testSuiteName flags pkg lbi = do
...
-      let additionalDirs' = maybe [] words
-            $ lookup "x-doctest-source-dirs"
-            $ customFieldsBI testBI
-      additionalDirs <- mapM (fmap ("-i" ++) . makeAbsolute) additionalDirs'
+           let additionalDirs' = maybe [] words
+                 $ lookup "x-doctest-source-dirs"
+                 $ customFieldsBI testBI
+           additionalDirs <- mapM (fmap ("-i" ++) . makeAbsolute) additionalDirs'
 
+           let compSuffix          = maybe "" ("_" ++) (mbCompName comp)
```
Comment (phadej) [inline src/Distribution/Extra/Doctest.hs]: component name may be invalid identifier, e.g in *can* contain dash, which isn't too far fetched.

Comment (RyanGlScott) [reply src/Distribution/Extra/Doctest.hs]: Good catch. I'm now applying the same `fixchar` trick that `Cabal` does when generating `MIN_VERSION_foo_bar` macros (for a package `foo-bar`).

**Comment by RyanGlScott** [comment]
I literally took `fixchar` from [the `Cabal` source code,](https://github.com/haskell/cabal/blob/20de0bfea72145ba1c37e3f500cee5258cc18e51/Cabal/Distribution/Simple/Build/Macros.hs#L156-L158) so if there's a problem with how `fixchar` is currently implemented, then there's a problem in `Cabal` too!

**Comment by phadej** [comment]
oh dear. Well, could you add a comment about `fixchar`?

**Comment by RyanGlScott** [comment]
Does [this](https://github.com/phadej/cabal-doctest/pull/28/files#diff-6cab482845abd89544c43b4491ec5c9bR367) suffice?

**Comment by phadej** [comment]
I'd like to have the link to the Cabal source next to fixchar, it will be sufficient.

**Comment by RyanGlScott** [comment]
@phadej, done.

**Comment by peti** [comment]
Ping? Have you had a chance to look at this PR?

**Comment by phadej** [comment]
I'll complete this today, thanks for reminding

**Pull request merged**